src/OrchardCore.Modules/OrchardCore.DataProtection.Azure/Startup.cs

```diff
@@ -0,0 +1,79 @@
+using System;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Modules;
+
+namespace OrchardCore.DataProtection.Azure
+{
+    public class Startup : StartupBase
+    {
+        private readonly IConfiguration _configuration;
+        private readonly IOptions<ShellOptions> _shellOptions;
+        private readonly ShellSettings _shellSettings;
+        private readonly ILogger<Startup> _logger;
+
+        public Startup(
+            IConfiguration configuration, 
+            IOptions<ShellOptions> shellOptions,
+            ShellSettings shellSettings, 
+            ILogger<Startup> logger)
+        {
+            _configuration = configuration;
+            _shellOptions = shellOptions;
+            _shellSettings = shellSettings;
+            _logger = logger;
+        }
+
+        public override void ConfigureServices(IServiceCollection services)
+        {
+            var blobName = $"{_shellOptions.Value.ShellsContainerName}/{_shellSettings.Name}/DataProtectionKeys.xml";
+
+            // Re-register the data protection services to be tenant-aware so that modules that internally
+            // rely on IDataProtector/IDataProtectionProvider automatically get an isolated instance that
+            // manages its own key ring and doesn't allow decrypting payloads encrypted by another tenant.
+            services.Add(new ServiceCollection()
```

Comment:
If this module depends on the main DP module, you'll be able to get rid of this hack here and just use:

```csharp
services.AddDataProtection().PersistKeysToAzureBlobStorage(GetBlobContainer(), blobName);
```